### PR TITLE
Adds respawn config option forcing respawn as another character slot

### PR DESCRIPTION
**Pull request by MrMelbert**
## About The Pull Request

Adds an option to the respawn config which forces you to pick another character (slot) before you respawn.

## Why It's Good For The Game

Just an idea i'm throwing out there, not necessarily pushing for it to be enabled on any servers. 

Respawning as an alternative character can be a good way to make people less frustrated at dying, particularly if paired with the cooldown config that already exists: 

"Oh shucks, I died and got my head cut off and got absorbed and got spaced by some changeling. I won't be able to finish my project or whatever. At least in 15 minutes I may be able to join as my botanist character to try something else rather than having to wait an hour and a half for the round to tick over." 

Also nice for downstream support. 

(Obviously you can just, *ban* people who respawn as the same character, use an honor system, but codifying it seems better than not.)

## Changelog

:cl: Melbert
config: Adds a config option for player respawning that enables respawns, but forces you pick a new character. 
config: "NORESPAWN" has been replaced with "ALLOW_RESPAWN 0". Unlimited respawns is "ALLOW_RESPAWN 1" and character limited respawns is "ALLOW_RESPAWN 2". 
/:cl:

### CODE REVIEW

**Review by Mothblocks — CHANGES_REQUESTED** [review]
you need to make sure existing NORESPAWN configs still work

**Comment by Mothblocks** [comment]
also your motive sucks because "reluctance to death" is extremely intentional

**Comment by MrMelbert** [comment]
> you need to make sure existing NORESPAWN configs still work

i'm not sure exactly what you mean, but the past default was "respawn off", and the new default is still "respawn off", so the only people who would need to update their config is people who had NORESPAWN commented out

**Comment by LemonInTheDark** [comment]
Not a massive fan of the WIGFTG either

**Comment by SimplySmiley123** [comment]
> Not a massive fan of the WIGFTG either. I think the issue we have currently is a culture that cannot accept things going "wrong", not people being upset in the first place.
> 
> I want dying to be impactful, but I also want players who are ok with that impact, or ok with their plans being foiled. That, IMO, is what allows for variable and fun stories while retaining a motivational link

Though let's be honest it still sucks pretty badly when you get removed from the round early on especially with nothing to do but close the game at that point and wait for the next round.
It's not just interesting stories believe me, there's a load of players who will just round remove you with no word said as fast as possible on TG.

**Comment by san7890** [comment]
dying should be impactful but i can see how such a framework is useful for downstreams or other servers that may not be as heavily focused on keeping players shut-out of the round

File: code/modules/mob/mob.dm [hunk 1]
```diff
@@ -768,18 +768,26 @@
...
-			return
-		else if (tgui_alert(usr, "Respawning is currently disabled, do you want to use your permissions to circumvent it?", "Respawn", list("Yes", "No")) != "Yes")
-			return
+	switch(CONFIG_GET(flag/allow_respawn))
+		if(RESPAWN_FLAG_NEW_CHARACTER)
+			if(tgui_alert(usr, "Note, respawning is only allowed as another character. If you don't have another free slot you may not be able to respawn.", "Respawn", list("Ok", "Nevermind")) != "Ok")
+				return
+
+		if(RESPAWN_FLAG_DISABLED)
+			// pass() // Normal respawn
```
Comment (san7890) [inline code/modules/mob/mob.dm:783]: no commented out code

**Review by san7890 — CHANGES_REQUESTED** [review]

File: code/modules/mob/mob.dm (hunk 1 above)
Comment (MrMelbert) [reply code/modules/mob/mob.dm:783]:
```suggestion
		if(RESPAWN_FLAG_DISABLED)
			pass() // Normal respawn
```

File: code/modules/admin/verbs/server.dm [hunk 2]
```diff
@@ -185,16 +185,42 @@
...
+		if(RESPAWN_FLAG_NEW_CHARACTER) // respawn currently enabled for different slot characters only
+			new_state = RESPAWN_FLAG_DISABLED
+			new_state_text = "Disabled"
+			to_chat(world, span_bold("You may no longer respawn :("), confidential = TRUE)
+
+		else
+			WARNING("Invalid respawn state in config: [respawn_state]")
+
+	if(new_state == -1)
+		to_chat(usr, span_warning("The config for respawn is set incorrectly, please complain to your nearest sever host (or fix it yourself). \
```
Comment (san7890) [inline code/modules/admin/verbs/server.dm:212]:
```suggestion
		to_chat(usr, span_warning("The config for respawn is set incorrectly, please complain to your nearest server host (or fix it yourself). \
```

File: code/modules/admin/verbs/server.dm [hunk 3]
```diff
@@ -185,16 +185,42 @@
...
+
+		if(RESPAWN_FLAG_FREE) // respawn currently enabled
+			new_state = RESPAWN_FLAG_NEW_CHARACTER
+			new_state_text = "Enabled, Different Slot"
+			to_chat(world, span_bold("You may now respawn as a different character."), confidential = TRUE)
+
+		if(RESPAWN_FLAG_NEW_CHARACTER) // respawn currently enabled for different slot characters only
+			new_state = RESPAWN_FLAG_DISABLED
+			new_state_text = "Disabled"
+			to_chat(world, span_bold("You may no longer respawn :("), confidential = TRUE)
```
Comment (san7890) [inline code/modules/admin/verbs/server.dm:206]: ehhhh i know it was here before we touched it but do you still think we should have the frowny face here or should we axe it?

File: code/modules/mob/mob.dm [hunk 4]
```diff
@@ -768,18 +768,26 @@
...
+				return
+
+		if(RESPAWN_FLAG_DISABLED)
+			pass() // Normal respawn
+
+		if(RESPAWN_FLAG_DISABLED)
+			if (!check_rights_for(usr.client, R_ADMIN))
+				to_chat(usr, span_boldnotice("Respawning is not enabled!"))
+				return
+			else if (tgui_alert(usr, "Respawning is currently disabled, do you want to use your permissions to circumvent it?", "Respawn", list("Yes", "No")) != "Yes")
```
Comment (san7890) [inline code/modules/mob/mob.dm:789]:
```suggestion
			if (tgui_alert(usr, "Respawning is currently disabled, do you want to use your permissions to circumvent it?", "Respawn", list("Yes", "No")) != "Yes")
```

we early return on the first anyways so i think we should drop the `else` for brevity now that we're having hella long tgui_alert lines

File: code/modules/admin/verbs/server.dm [hunk 5]
```diff
@@ -185,16 +185,42 @@
...
+
+		else
+			WARNING("Invalid respawn state in config: [respawn_state]")
+
+	if(new_state == -1)
+		to_chat(usr, span_warning("The config for respawn is set incorrectly, please complain to your nearest sever host (or fix it yourself). \
+			In the meanwhile respawn has been sets to \"Off\"."))
```
Comment (san7890) [inline code/modules/admin/verbs/server.dm:213]:
```suggestion
			In the meanwhile respawn has been set to \"Off\"."))
```

**Review by san7890 — CHANGES_REQUESTED** [review]
some more nits, few are grammar and others are stuff that was just upheaved in the midst of larger reworks

File: code/modules/admin/verbs/server.dm (hunk 3 above)
Comment (MrMelbert) [reply code/modules/admin/verbs/server.dm:206]: It's SOUL i'm not removing it

Comment (san7890) [reply code/modules/admin/verbs/server.dm:206]: alright just soliciting your opinion here. i wish we could have punctuation here though (insert xkcd comic here)

**Pull request merged**